### ISSUE TITLE
Release Branch coverage

### ISSUE THREAD

**Issue by PragTob** [colszowka/simplecov#781]
After merging #694 this issue is here to keep track of the functionality until initial release and helps me & others :)

Big thanks to @som4ik @tycooon @stepozer @klyonrad and many others for their valuable contributions!

* [x] Merge initial PR #694 
* [x] Merge corresponding html formatter PR colszowka/simplecov-html#64
  * [x] Address remaining code review feedback
* [x] Make sure problems with existing PR are solved ([ref1](https://github.com/colszowka/simplecov/pull/694#issuecomment-561092374) [ref2](https://github.com/colszowka/simplecov/pull/694#issuecomment-561095694))
  * [x] parallel tests support works as intended
  * [x] `:nocov:` blocks should be considered
* [x] Address [initial code review comments](https://github.com/colszowka/simplecov/pull/694#pullrequestreview-256178924)
* [x] Change [used option name](https://github.com/colszowka/simplecov/pull/694#discussion_r298918548) to use branch coverage/chanege coverage type
* [x] add some feature tests
  * [x] for instance with the neat [tester script](https://github.com/colszowka/simplecov/pull/694#issuecomment-562097006)
* [x] Investigate coverage reporting (it seems like it reports lines as covered when what we'd like to have are branches)
* [x] release pre release version for general/public testing (this and formatter)
* [ ] make it possible that the branch coverage percentage is displayed as the "main" coverage and all our checks take it instead of line coverage
* [x] also allow people to do coverage checks versus multiple metrics for instance I want my line coverage to be at least 95% and my branch coverage to be at least 90%.
* [x] figure out problems with the format data is stored in #781 
* [x] fix reported beta problems
* [x] release the "finished for now" branch coverage

**Comment by PragTob** [comment]
:wave: 

Hi @som4ik - got around to writing a feature test for this. It seems to me (and my reading of the code) that we still report coverage on a per line basis - is this correct from your memory? :sweat_smile: 

What I mean by that is that we use branch coverage to figure out which lines are covered and then we mark those lines as covered and report them as such. I think what we want is that we report it on a per branch basis (so instead of 80/100 lines covered we wanna say 33/38 branches covered or something to the tune of this). 

Confirmation would be nice!

**Comment by som4ik** [comment]
Hi @PragTob :wave: 
As I remember per line bases report is mandatory-default coverage report, passing `:branches` parameter to `Coverage.start` adds a new type of collected coverage result matrics, but doesn't cancel generating per line bases report.
Just to clarify, we don't use per line bases to generate branch coverage metrics, you can notice that the coverage report displays already your suggestion: 
![branch-1](https://user-images.githubusercontent.com/7699543/70886636-fcaa8d00-1fec-11ea-9488-126c0b3323ea.png)
![branch-2](https://user-images.githubusercontent.com/7699543/70886648-03390480-1fed-11ea-9e07-897d1969703a.png)
let me guess what made it confusing: 
- misleading here is the config parameter naming, `use_branchable_report` it should be at least `add_branchable_report`.
- Branch class contains matrics start_line, end_line. (it's branch coverage metrics and it's not related to the  per lines bases report ).
Hope I answered the question :smile:

**Comment by PragTob** [comment]
Yeah you did thank you, then I just gotta tweak the cukes and selects on all of that. Thank you!

**Comment by PragTob** [comment]
Yeah so I took a look at it and figured out that right now we basically just "add" branch coverage but there's no way to make it the "main coverage" aka the coverage by we fail if we say minimum coverage or something else.

I think I want to do a first release with that, but I think it'd be good if people could decide what their "main coverage" would be (and then also display that most prominently).

Another thing that my tests just showed me as counter intuitive is that if a file has no conditional statements at all there are no branches and hence we report 0 branches and say it's 100% covered even if no method defined in there is ever covered. Which is plenty weird I believe. On the other hand branch coverage by itself can't determine whether it's covered or not as the coverage data is literally empty...

**Comment by JonJagger** [comment]
I would have thought that a file with no content at all had zero branches... 
A file with content but no ifs/ternarys etc has one branch, which is the 'main' branch...
$0.02
P.S. Awesome work. Can't wait to add this to my numerous Ruby repos :)

**Comment by JonJagger** [comment]
Some code that previously had 100% coverage just dropped to 99% coverage.
Looking at the html report it shows this:
<img width="1159" alt="simplecov-screenshot" src="https://user-images.githubusercontent.com/252118/71556741-0512b700-2a34-11ea-86c8-eb6fda6d0758.png">
I don't understand how line 42 can have no coverage when it is a `while` and the block it controls 43-46 does have coverage (24 hits). My only guess is it's something to do with the previous line, 41, which has an && in it?

**Comment by PragTob** [comment]
@JonJagger :wave: 

That seems unlikely, doesn't look like you have branch coverage activated (there would be little badges saying "+", 0 or something around there or something). 

I absolutely agree the line should be covered though, so that seems odd. But it reminds me off #764 where it was a coverage bug that some until line wasn't covered. 

Are you sure that this is related to some simplecov upgrade or maybe some ruby version upgrade in the middle of it?

**Comment by tleish** [comment]
@JonJagger - out of curiosity, what does the coverage for `def same?` look like?

**Comment by JonJagger** [comment]
The coverage for `def same` is
<img width="1184" alt="simplecov-same-coverage" src="https://user-images.githubusercontent.com/252118/71557890-735f7580-2a44-11ea-82f0-60374075d415.png">

**Comment by JonJagger** [comment]
> @JonJagger 👋
> 
> That seems unlikely, doesn't look like you have branch coverage activated (there would be little badges saying "+", 0 or something around there or something).
>
 
Yes. Sorry. I was not clear. My assumption was that my recent upgrade had got a new version of simplecov which had been refactored in preparation for the branch-coverage.

> I absolutely agree the line should be covered though, so that seems odd. But it reminds me off #764 where it was a coverage bug that some until line wasn't covered.
> 
> Are you sure that this is related to some simplecov upgrade or maybe some ruby version upgrade in the middle of it?

The app in question is docker containerized. It gets simplecov and ruby from a base image (Alpine).
Shelling into that I can see the Gemfile.lock
```
GEM
  remote: https://rubygems.org/
  specs:
    docile (1.3.2)
    json (2.3.0)
    minitest (5.13.0)
    ruby-prof (1.1.0)
    simplecov (0.17.1)
      docile (~> 1.1)
      json (>= 1.8, < 3)
      simplecov-html (~> 0.10.0)
    simplecov-html (0.10.2)
```
And ruby -v gives
```
ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-linux-musl]
```

The previous docker image gives...
```
GEM
  remote: https://rubygems.org/
  specs:
    daemons (1.3.1)
    docile (1.3.2)
    eventmachine (1.2.7)
    json (2.2.0)
    minitest (5.11.3)
    prometheus-client (0.9.0)
      quantile (~> 0.2.1)
    quantile (0.2.1)
    rack (2.0.7)
    ruby-prof (0.18.0)
    simplecov (0.16.1)
      docile (~> 1.1)
      json (>= 1.8, < 3)
      simplecov-html (~> 0.10.0)
    simplecov-html (0.10.2)
    thin (1.7.2)
      daemons (~> 1.0, >= 1.0.9)
      eventmachine (~> 1.0, >= 1.0.4)
      rack (>= 1, < 3)
```
and
```
ruby 2.5.5p157 (2019-03-15 revision 67260) [x86_64-linux-musl]
```

The code being covered has not changed.

**Comment by JonJagger** [comment]
Rebuilding with simplecov set to 0.16.1 the coverage anomaly goes away...
<img width="1186" alt="simplecov-0 16 1-coverage" src="https://user-images.githubusercontent.com/252118/71558217-8c6a2580-2a48-11ea-82ae-018350449cb3.png">
I note also that ruby -v gives 2.5.7p206

**Comment by PragTob** [comment]
@JonJagger could you please open a separate issue as it seems to be not connected to this then? Also the ruby version changes so that's still my suspect but it could ofc still be that 0.17.0/0.17.1 broke something.

**Comment by JonJagger** [comment]
I am now unable to reproduce the anomaly with simplecov set to 0.16.1 and 0.17.0 and 0.17.1 
The difference is this time I created the images from scratch without relying on possibly stale image caches. Viz, user error. Apologies for the false alarm.

**Comment by PragTob** [comment]
@JonJagger maybe running tests in parallel? That makes things a lot fancier and more likely to have weird errors :)

**Comment by JonJagger** [comment]
The problem is the base OS version. 
When its Alpine:3.10 (simplecov=0.17.1, ruby -v == 2.5.7p206) then no anomaly.
When its Alpine:3.11 (simplecov=0.17.1, ruby -v == 2.6.5p114) then I get the anomaly.
I'll open a separate issue.

**Comment by PragTob** [comment]
Please enjoy the pre-release of 0.18.0.beta1 - I'd be happy if you tested it out. Specifically if parallel tests and merging works, I haven't tried it out myself yet because if I'm gonna set all of that up then I might as well integrate it into the test suite (finally!!!!) and that requires some more work on cucumber.

**Comment by dmitry** [comment]
simplecov-html incorrectly sorting percentages, using string instead of float values, because of that first comes the numbers with less chars.

![Desktop screenshot](https://user-images.githubusercontent.com/5091/71975016-49ac0880-3213-11ea-80b3-d1c0cca05c14.png)

**Comment by PragTob** [comment]
Hi @dmitry 

thanks, but that's most likely not related to branch coverage and is likely a long standing bug with the JS in the html report.

Created https://github.com/colszowka/simplecov-html/issues/79 for it.

**Comment by klyonrad** [comment]
Sadly I have to report a breaking bug 😿 

```
bundle show simplecov
/Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1

bundle show simplecov-html
/Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-html-0.11.0.beta1

bundle exec rspec spec/policies
Run options: exclude {:future=>true}                                                                                 

Randomized with seed 25469
 237/237 |========================================== 100 ==========================================>| Time: 00:00:23 

Finished in 23.37 seconds (files took 10.5 seconds to load)
237 examples, 0 failures

Randomized with seed 25469
Traceback (most recent call last):
	15: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/defaults.rb:29:in `block in <top (required)>'
	14: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov.rb:164:in `run_exit_tasks!'
	13: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/configuration.rb:196:in `block in at_exit'
	12: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov.rb:74:in `result'
	11: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result_merger.rb:83:in `merged_result'
	10: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result_merger.rb:57:in `results'
	 9: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result_merger.rb:57:in `each'
	 8: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result_merger.rb:58:in `block in results'
	 7: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result.rb:72:in `from_hash'
	 6: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result.rb:86:in `symbolize_names_of_coverage_results'
	 5: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result.rb:86:in `each_with_object'
	 4: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result.rb:86:in `each'
	 3: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result.rb:87:in `block in symbolize_names_of_coverage_results'
	 2: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result.rb:87:in `each_with_object'
	 1: from /Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result.rb:87:in `each'
/Users/username/.rvm/gems/ruby-2.6.5/gems/simplecov-0.18.0.beta1/lib/simplecov/result.rb:88:in `block (2 levels) in symbolize_names_of_coverage_results': undefined method `to_sym' for 1:Integer (NoMethodError)
Did you mean?  to_s

```


Our simplecov setup looks like this

```ruby



# spec_helper

ENV['RAILS_ENV'] ||= 'test'

if ENV['RAILS_ENV'] == 'test' && !File.exist?('.rspec-disable-simplecov')
  require 'simplecov'
  SimpleCov.start 'rails' do
    enable_coverage :branch
    add_group 'DataTables', 'app/datatables'
    add_group 'Decorators', 'app/decorators'
    add_group 'Finders', 'app/finders'
    add_group 'PDFs', 'app/pdfs'
    add_group 'Excel', 'app/xlsx'
    add_group 'Policies', 'app/policies'
    add_group 'Serializers', 'app/serializers'
    add_group 'ServiceObjects', 'app/service_objects'
    add_group 'Uploaders', 'app/uploaders'
    add_group 'Validators', 'app/validators'
  end

  SimpleCov.formatter = SimpleCov::Formatter::HTMLFormatter

  if ENV['CODECOV_TOKEN']
    require 'codecov'
    SimpleCov.formatter = SimpleCov::Formatter::Codecov
  end
end

# Gemfile

group :development, :test do
  # ...
  gem 'simplecov', '0.18.0.beta1'
end

group :test do
  gem 'codecov'
  # ....
end

```

**Comment by deivid-rodriguez** [comment]
I run into the same thing yesterday, was trying to find time to setup a clean repo and maybe fix it myself.

**Comment by PragTob** [comment]
Thanks for testing :green_heart: ! Yeah that is result merging... test coverage is spotty there, one of the reasons I'm currently trying to up the number of features and number of projects we can support.

You're probably running either parallel tests or two different test suites, right? 

Code looks odd at the moment, it deals with the coverage data but doesn't seem to deal with lines vs branches coverage which seems odd given that both should be present. That part of the code, I'm also sadly not very familiar with.

Also the JSON.parse call is now using `symbolize_names: true` - so not too sure if that code is actually needed anymore :man_shrugging:

**Comment by klyonrad** [comment]
> You're probably running either parallel tests or two different test suites, right?

Me, no. We even uninstalled the parral testing gem recently (sorry, forgot to mention that)

**Comment by PragTob** [comment]
@klyonrad wait what like... just one rspec test suite or what? I thought it was rspec + cucumber because the code is the result merger code... Guess I gotta really dig into this not on the side while my head is a bit dizzy. Then I can see under which circumstances this code is triggered. 

(my vague suspicion right now is that an old "resultset.json" or something might be lying around with the old data format and that's what breaks it but :man_shrugging: )

**Comment by deivid-rodriguez** [comment]
@PragTob Note that simplecov automerges new coverage reports with old ones, so the code can be triggered by a simple setup I think?

Thanks for all your work, by the way :purple_heart:.

**Comment by klyonrad** [comment]
> Note that simplecov automerges new coverage reports with old ones

What the hell 😱

So uhm... I deleted the `coverage/` directory before running and this seems to cause the report generation to hang for quite while (and I needed to force interrupt)

**Comment by klyonrad** [comment]
## Guard clauses and nested conditionals in one line

![image](https://user-images.githubusercontent.com/1453955/72090720-e3a4ab80-330e-11ea-8a10-cb7d369edef1.png)

Shouldn't this be three branches?

## two sided and

![image](https://user-images.githubusercontent.com/1453955/72090856-24042980-330f-11ea-84d0-372aec8db788.png)

Isn't this two branches?

**Comment by tycooon** [comment]
> Shouldn't this be three branches?

Unfortunately, this is not how Ruby branch coverage work. You could store the condition result in a variable `x` and your code would look like `return if x`. So this is just 2 branches from ruby perspective. 

In case you care about branch coverage a lot, you should probably structure your code like so:

```ruby
return if cond1
return if cond2
true
```

This way you can ensure that all combinations are covered.

**Comment by tycooon** [comment]
> Isn't this two branches?
No, because there are no conditionals in this code. Method `any` might be always returning `42`, who knows? Ruby is a dynamic language.

**Comment by PragTob** [comment]
@klyonrad yeah what @tycooon said :) Seems odd to me that the line is makred red though, I'd expect it to be green since it has line coverage (but missing branch coverage) - haven't looked into that yet but I might like it better this way either way :D

**Comment by PragTob** [comment]
@klyonrad @deivid-rodriguez moved that issue to #820 and will try to chip away at it today as it's a rather critical bug.

**Comment by PragTob** [comment]
0.18.0.beta2 has been released, I'd be happy if y'all gave it another should have some really great improvements! https://github.com/colszowka/simplecov/blob/master/CHANGELOG.md#0180beta2-2020-01-19

**Comment by PragTob** [comment]
has anyone tried out the beta2/beta3 releases? Would be happy about feedback!

**Comment by klyonrad** [comment]
Not on a grand scale, only sometimes: It does not break 🎉 😍 

The change in showing names and hit counts (!!) for the branch certainly makes it better. Although it starts to look funny when you have a lot (!) of branches in one line 🔱

Even in string interpolation things you can see the difference, or even more subtle (and mean) the nil check (`&.`) operator!

I will try to motivate my former team to test it in their project and report success or failure (the one where I tested over a year and reported issues)

**Comment by PragTob** [comment]
@klyonrad thanks a bunch!

I might just go ahead and release it though (testing is great anyhow) but I figure that's the best way to get people to test it. 

I haven't even tested it with `&.` - probably should add a test case for that just for fun and profit.

If you have ideas how to make multiple per one line more readable, your ideas are welcome. Off the top of my I'd say sort them by order of appearance (we have the start_col/end_col available) or maybe show more information on hover.

Thanks!

**Comment by PragTob** [comment]
And 0.18.0 is released :tada: 

Didn't manage to get in the display as "main coverage" but I don't think it's the main killer feature. There are some spots where we need to up support and they're in my head and I might/will make issues for them.

I figured it was more important to finally get this into everyone's hands so they can let me/us know what works and what doesn't.
